README.rst
dataclass-builder
=================

|build-status|
|doc-status|
|coverage-status|
|code-style|
|version|
|supported-versions|
|status|
|license|

Create instances of Python dataclasses with the builder pattern.


Documentation
-------------

Documentation for PyRADS can be found at `https://dataclass-builder.readthedocs.io/en/latest/ <https://dataclass-builder.readthedocs.io/en/latest/>`_ or in `PDF <https://readthedocs.org/projects/dataclass-builder/downloads/pdf/latest/>`_ and `Epub <https://readthedocs.org/projects/dataclass-builder/downloads/epub/latest/>`_ formats.


Requirements
------------

* Python 3.6 or greater
* dataclasses_ if using Python 3.6


Installation
------------

`dataclass-builder` is on PyPI_ so the easiest way to install it is:

.. code-block:: text

    $ pip install dataclass-builder


Usage
-----

There are two ways to use `dataclass-builder`.  Via a builder instance or by creating a dedicated builder.  The latter is recommended when repeated building of a given dataclass is desired or when docstrings and type checking are important.


Dedicated Builder (builder factory)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Using specialized builders allows for better documentation than the `DataclassBuilder` wrapper.

.. code-block:: python

    from dataclasses import dataclass
    from dataclass_builder import (dataclass_builder, build, fields, update
                                   REQUIRED, OPTIONAL)

    @dataclass
    class Point:
        x: float
        y: float
        w: float = 1.0

    PointBuilder = dataclass_builder(Point)

Now we can build a point.

.. code-block:: python

    >>> builder = PointBuilder()
    >>> builder.x = 5.8
    >>> builder.y = 8.1
    >>> builder.w = 2.0
    >>> build(builder)
    Point(x=5.8, y=8.1, w=2.0)

As long as the dataclass_ the builder was constructed for does not have a `build` field then a `build` method will be generated as well.

    >>> builder.build()
    Point(x=5.8, y=8.1, w=2.0)

Field values can also be provided in the constructor.

.. code-block:: python

    >>> builder = PointBuilder(x=5.8, w=100)
    >>> builder.y = 8.1
    >>> builder.build()
    Point(x=5.8, y=8.1, w=100)

*Positional arguments are not allowed.*

Fields with default values in the dataclass_ are optional in the builder.

.. code-block:: python

    >>> builder = PointBuilder()
    >>> builder.x = 5.8
    >>> builder.y = 8.1
    >>> builder.build()
    Point(x=5.8, y=8.1, w=1.0)

Fields that don't have default values in the dataclass_ are not optional.

.. code-block:: python

    >>> builder = PointBuilder()
    >>> builder.y = 8.1
    >>> builder.build()
    Traceback (most recent call last):
    ...
    MissingFieldError: field 'x' of dataclass 'Point' is not optional

Fields not defined in the dataclass cannot be set in the builder.

.. code-block:: python

    >>> builder.z = 3.0
    Traceback (most recent call last):
    ...
    UndefinedFieldError: dataclass 'Point' does not define field 'z'

*No exception will be raised for fields beginning with an underscore as they are reserved for use by subclasses.*

Accessing a field of the builder before it is set gives either the `REQUIRED` or `OPTIONAL` constant

.. code-block:: python

    >>> builder = PointBuilder()
    >>> builder.x
    REQUIRED
    >>> builder.w
    OPTIONAL

The `fields` method can be used to retrieve a dictionary of settable fields for the builder.  This is a mapping of field names to `dataclasses.Field` objects from which extra data can be retrieved such as the type of the data stored in the field.

.. code-block:: python

    >>> list(builder.fields().keys())
    ['x', 'y', 'w']
    >>> [f.type.__name__ for f in builder.fields().values()]
    ['float', 'float', 'float']

A subset of the fields can be also be retrieved, for instance, to only get required fields:

.. code-block:: python

    >>> list(builder.fields(optional=False).keys())
    ['x', 'y']

or only the optional fields.

.. code-block:: python

    >>> list(builder.fields(required=False).keys())
    ['w']


If the underlying dataclass_ has a field named `fields` this method will not be generated and instead the `fields` function should be used instead.

An already built dataclass_ can be updated with a partially completed builder using the :code:`update` function.

.. code-block:: python

    >>> point = Point(x=5.8, y=8.1, w=100)
    >>> update(point, PointBuilder(y=1.1))
    >>> point
    Point(x=5.8, y=1.1, w=100)

*Dataclass builders can also be updated, but frozen dataclasses cannot.*


Builder Instance (generic wrapper)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Using a builder instance is the fastest way to get started with the `dataclass-builder` package.

.. code-block:: python

    from dataclasses import dataclass
    from dataclass_builder import (DataclassBuilder, build, fields,
                                   REQUIRED, OPTIONAL)

    @dataclass
    class Point:
        x: float
        y: float
        w: float = 1.0

Now we can build a point.

.. code-block:: python

    >>> builder = DataclassBuilder(Point)
    >>> builder.x = 5.8
    >>> builder.y = 8.1
    >>> builder.w = 2.0
    >>> build(builder)
    Point(x=5.8, y=8.1, w=2.0)

Field values can also be provided in the constructor.

.. code-block:: python

    >>> builder = DataclassBuilder(Point, x=5.8, w=100)
    >>> builder.y = 8.1
    >>> build(builder)
    Point(x=5.8, y=8.1, w=100)

.. note::

    Positional arguments are not allowed, except for the dataclass_ itself.

Fields with default values in the dataclass_ are optional in the builder.

.. code-block:: python

    >>> builder = DataclassBuilder(Point)
    >>> builder.x = 5.8
    >>> builder.y = 8.1
    >>> build(builder)
    Point(x=5.8, y=8.1, w=1.0)

Fields that don't have default values in the dataclass_ are not optional.

.. code-block:: python

    >>> builder = DataclassBuilder(Point)
    >>> builder.y = 8.1
    >>> build(builder)
    Traceback (most recent call last):
    ...
    MissingFieldError: field 'x' of dataclass 'Point' is not optional

Fields not defined in the dataclass cannot be set in the builder.

.. code-block:: python

    >>> builder.z = 3.0
    Traceback (most recent call last):
    ...
    UndefinedFieldError: dataclass 'Point' does not define field 'z'

.. note::

    No exception will be raised for fields beginning with an underscore as they are reserved for use by subclasses.

Accessing a field of the builder before it is set gives either the `REQUIRED` or `OPTIONAL` constant

.. code-block:: python

    >>> builder = DataclassBuilder(Point)
    >>> builder.x
    REQUIRED
    >>> builder.w
    OPTIONAL

The `fields` function can be used to retrieve a dictionary of settable fields for the builder.  This is a mapping of field names to `dataclasses.Field` objects from which extra data can be retrieved such as the type of the data stored in the field.

.. code-block:: python

    >>> list(fields(builder).keys())
    ['x', 'y', 'w']
    >>> [f.type.__name__ for f in fields(builder).values()]
    ['float', 'float', 'float']

A subset of the fields can be also be retrieved, for instance, to only get required fields:

.. code-block:: python

    >>> list(fields(builder, optional=False).keys())
    ['x', 'y']

or only the optional fields.

.. code-block:: python

    >>> list(fields(builder, required=False).keys())
    ['w']



.. _dataclass: https://github.com/ericvsmith/dataclasses
.. _dataclasses: https://github.com/ericvsmith/dataclasses
.. _PyPI: https://pypi.org/

.. |build-status| image:: https://travis-ci.com/mrshannon/dataclass-builder.svg?branch=master&style=flat
    :target: https://travis-ci.com/mrshannon/dataclass-builder
    :alt: Build status

.. |doc-status| image:: https://readthedocs.org/projects/dataclass-builder/badge/?version=latest
   :target: https://dataclass-builder.readthedocs.io/en/latest/
   :alt: Documentation status

.. |coverage-status| image:: https://codecov.io/gh/mrshannon/dataclass-builder/coverage.svg?branch=master
    :target: https://codecov.io/gh/mrshannon/dataclass-builder?branch=master
    :alt: Test coverage

.. |code-style| image:: https://img.shields.io/badge/code%20style-black-000000.svg
    :target: https://github.com/psf/black
    :alt: Code style is black

.. |version| image:: https://img.shields.io/pypi/v/dataclass-builder.svg
    :alt: PyPI Package latest release
    :target: https://pypi.python.org/pypi/dataclass-builder

.. |status| image:: https://img.shields.io/pypi/status/dataclass-builder.svg
    :alt: Status
    :target: https://pypi.python.org/pypi/dataclass-builder

.. |supported-versions| image:: https://img.shields.io/pypi/pyversions/dataclass-builder.svg
    :alt: Supported versions
    :target: https://pypi.python.org/pypi/dataclass-builder

.. |supported-implementations| image:: https://img.shields.io/pypi/implementation/dataclass-builder.svg
    :alt: Supported implementations
    :target: https://pypi.python.org/pypi/dataclass-builder

.. |license| image:: https://img.shields.io/github/license/mrshannon/dataclass-builder.svg
    :alt: MIT
    :target: https://opensource.org/licenses/MIT



CHANGELOG.rst
Changelog
=========


Unreleased_
-----------


v1.2.0_ - 2019-08-21
--------------------

* Add :code:`update` function to facilitate updating a dataclass or dataclass
  builder from a partially completed dataclass builder.


v1.1.3_ - 2019-08-07
--------------------

* Switch to reST docstrings and remove paramref requirement.
* Fix bug in dataclass builders that contains a field named fields.
* Fix return type annotation of the build method.
* Fix issue `#12`_ which would cause sentinel values :code:`REQUIRED`,
  :code:`OPTIONAL`, and :code:`MISSING` to become incomparable after
  a :code:`copy` or :code:`deepcopy`.


v1.1.2_ - 2019-03-29
--------------------

* Fix issue `#9`_ relating to the :code:`dataclass_builder` factory function
  failing to handle dataclasses which use types from the :code:`typing` module.
* Replace equality checks with identity checks where possible.  This was not
  only the intention of the checks but by using identity, compatibility with
  types that have broken equality operators is restored.


v1.1.1_ - 2019-03-27
--------------------

* Fix bug where the :code:`MISSING` constant was not exported at the package
  level.


v1.1.0_ - 2019-03-27
--------------------

* Make :code:`MISSING` constant comparable (True) with :code:`REQUIRED` and
  :code:`OPTIONAL` constants.


v1.0.1_ - 2019-03-26
--------------------

* Fixed trove classifiers.


v1.0.0_ - 2019-03-26
--------------------

*First stable release.*

**BREAKS BACKWARDS COMPATIBILITY**

* Moved :code:`REQUIRED` and :code:`OPTIONAL` constants into :code:`._common`
  with exports at the package level.
* Fixed bug in :code:`DataclassBuilder.__repr__` causing some types of fields
  such as strings to not display properly.
* Rewrite :code:`DataclassBuilder` to share functions with the factory
  version :code:`dataclass_builder`.
* Change how :code:`DataclassBuilder` handles unset fields so that they will
  be initialized to :code:`REQUIRED` and :code:`OPTIONAL` in order to be
  compatible with the classes returned from the :code:`dataclass_builder`
  factory.
* Fields passed in the :code:`__init__` method of :code:`DataclassBuilder` will
  now raise :code:`TypeError` (instead of :code:`UndefinedFieldError`) the same
  as classes build with the :code:`dataclass_builder` factory.
* Disabled positional arguments for builders generated with the
  :code:`dataclass_builder` factory.
* Fixed documentation mismatches.  The documentation both in the README and in
  the module docstrings properly reflect the API.


v0.0.3_ - 2019-03-26
--------------------

* Added :code:`.factory.dataclass_builder` factory function to create builder
  classes for specific dataclasses.
* Moved :code:`build` and :code:`fields` functions to the :code:`.utility`
  submodule with exports at the top level.
* Moved :code:`DataclassBuilder` to the :code:`.wrapper` submodule with exports
  at the top level.
* Fixed issue #4 which caused typing errors in code using
  :code:`DataclassBuilder`.


v0.0.2_ - 2019-03-12
--------------------

* Added additional unit tests.
* Fixed issue #1 preventing :code:`DataclassBuilder` from being subclassed.
* Fields beginning with an underscore will no longer be checked.


v0.0.1 - 2019-03-11
-------------------

* Added :code:`DataclassBuilder` class to wrap Python dataclasses_ with a
  factory implementing the builder pattern.
* Added :code:`build` function to construct a dataclass_ from a
  :code:`DataclassBuilder`.
* Added :code:`fields` function to extract information about the available
  fields from a :code:`DataclassBuilder`.


.. _dataclasses: https://docs.python.org/3/library/dataclasses.html
.. _dataclass: https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass

.. _Unreleased: https://github.com/mrshannon/dataclass-builder/compare/v1.2.0...HEAD
.. _v1.2.0: https://github.com/mrshannon/dataclass-builder/compare/v1.1.3...v1.2.0
.. _v1.1.3: https://github.com/mrshannon/dataclass-builder/compare/v1.1.2...v1.1.3
.. _v1.1.2: https://github.com/mrshannon/dataclass-builder/compare/v1.1.1...v1.1.2
.. _v1.1.1: https://github.com/mrshannon/dataclass-builder/compare/v1.1.0...v1.1.1
.. _v1.1.0: https://github.com/mrshannon/dataclass-builder/compare/v1.0.1...v1.1.0
.. _v1.0.1: https://github.com/mrshannon/dataclass-builder/compare/v1.0.0...v1.0.1
.. _v1.0.0: https://github.com/mrshannon/dataclass-builder/compare/v0.0.3...v1.0.0
.. _v0.0.3: https://github.com/mrshannon/dataclass-builder/compare/v0.0.2...v0.0.3
.. _v0.0.2: https://github.com/mrshannon/dataclass-builder/compare/v0.0.1...v0.0.2

.. _#12: https://github.com/mrshannon/dataclass-builder/issues/12
.. _#9: https://github.com/mrshannon/dataclass-builder/issues/9


dataclass_builder/factory.py
"""Create :func:`dataclasses.dataclass` builders for specific dataclasses.

This module uses a factory to build builder classes that build a specific
dataclass.  These builder classes implement the builder pattern and allow
constructing dataclasses over a period of time instead of all at once.


Examples
--------
Using specialized builders allows for better documentation than the
:class:`DataclassBuilder` wrapper and allows for type checking because
annotations are dynamically generated.

.. testcode::

    from dataclasses import dataclass
    from dataclass_builder import (dataclass_builder, build, fields,
                                   REQUIRED, OPTIONAL)

    @dataclass
    class Point:
        x: float
        y: float
        w: float = 1.0

    PointBuilder = dataclass_builder(Point)

Now we can build a point.

.. doctest::

    >>> builder = PointBuilder()
    >>> builder.x = 5.8
    >>> builder.y = 8.1
    >>> builder.w = 2.0
    >>> build(builder)
    Point(x=5.8, y=8.1, w=2.0)

As long as the dataclass the builder was constructed for does not have a
`build` field then a `build` method will be generated as well.

    >>> builder.build()
    Point(x=5.8, y=8.1, w=2.0)

Field values can also be provided in the constructor.

.. doctest::

    >>> builder = PointBuilder(x=5.8, w=100)
    >>> builder.y = 8.1
    >>> builder.build()
    Point(x=5.8, y=8.1, w=100)

.. note::

    Positional arguments are not allowed.

Fields with default values in the dataclass are optional in the builder.

.. doctest::

    >>> builder = PointBuilder()
    >>> builder.x = 5.8
    >>> builder.y = 8.1
    >>> builder.build()
    Point(x=5.8, y=8.1, w=1.0)

Fields that don't have default values in the dataclass are not optional.

.. doctest::

    >>> builder = PointBuilder()
    >>> builder.y = 8.1
    >>> builder.build()
    Traceback (most recent call last):
    ...
    MissingFieldError: field 'x' of dataclass 'Point' is not optional

Fields not defined in the dataclass cannot be set in the builder.

.. doctest::

    >>> builder.z = 3.0
    Traceback (most recent call last):
    ...
    UndefinedFieldError: dataclass 'Point' does not define field 'z'

.. note::

    No exception will be raised for fields beginning with an underscore as they
    are reserved for use by subclasses.

Accessing a field of the builder before it is set gives either the `REQUIRED`
or `OPTIONAL` constant

.. doctest::

    >>> builder = PointBuilder()
    >>> builder.x
    REQUIRED
    >>> builder.w
    OPTIONAL

The `fields` method can be used to retrieve a dictionary of settable fields for
the builder.  This is a mapping of field names to :class:`dataclasses.Field`
objects from which extra data can be retrieved such as the type of the data
stored in the field.

.. doctest::

    >>> list(builder.fields().keys())
    ['x', 'y', 'w']
    >>> [f.type.__name__ for f in builder.fields().values()]
    ['float', 'float', 'float']

A subset of the fields can be also be retrieved, for instance, to only get
required fields:

.. doctest::

    >>> list(builder.fields(optional=False).keys())
    ['x', 'y']

or only the optional fields.

.. doctest::

    >>> list(builder.fields(required=False).keys())
    ['w']

.. note::

    If the underlying dataclass has a field named `fields` this method will
    not be generated and instead the :func:`fields` function should be used
    instead.

"""

from typing import (
    TYPE_CHECKING,
    Any,
    Callable,
    Dict,
    Mapping,
    MutableMapping,
    Optional,
    Sequence,
    Type,
    cast,
)

from ._common import (
    MISSING,
    OPTIONAL,
    REQUIRED,
    _is_required,
    _optional_fields,
    _required_fields,
    _settable_fields,
)
from .exceptions import MissingFieldError, UndefinedFieldError

if TYPE_CHECKING:
    from dataclasses import Field, is_dataclass
else:
    from dataclasses import is_dataclass

__all__ = ["dataclass_builder"]


# copied (and modified) from dataclasses._create_fn to avoid dependency on
# private functions in dataclasses
def _create_fn(
    name: str,
    args: Sequence[str],
    body: Sequence[str],
    env: Optional[Dict[str, Any]] = None,
    *,
    return_type: Any = MISSING,
) -> Callable[..., Any]:
    locals_: MutableMapping[str, Any] = {}
    return_annotation = ""
    if env is None:
        env = {}
    if return_type is not MISSING:
        env["_return_type"] = return_type
        return_annotation = "->_return_type"
    args = ", ".join(args)
    body = "\n".join(f" {line}" for line in body)
    txt = f"def {name}({args}){return_annotation}:\n{body}"
    # this is how the dataclasses module makes custom methods so it's good
    # enough for this package
    exec(txt, env, locals_)  # pylint: disable=exec-used
    return cast(Callable[..., Any], locals_[name])


def _create_init_method(fields: Mapping[str, "Field[Any]"]) -> Callable[..., None]:
    env: Dict[str, Any] = {
        f"_{name}_type": field.type for name, field in fields.items()
    }
    env["REQUIRED"] = REQUIRED
    env["OPTIONAL"] = OPTIONAL

    def is_required(field: "Field[Any]") -> str:
        return "REQUIRED" if _is_required(field) else "OPTIONAL"

    if fields:
        args = ["self", "*"] + [
            f"{name}: _{name}_type = {is_required(field)}"
            for name, field in fields.items()
        ]
    else:
        args = ["self"]
    body = [f"self.{name}: _{name}_type = {name}" for name in fields]
    body = ["self.__initialized = False"] + body + ["self.__initialized = True"]
    return _create_fn("__init__", args, body, env, return_type=None)


def _create_class_docstring(dataclass: Any) -> str:
    dname = dataclass.__qualname__
    try:
        dname = dataclass.__module__ + "." + dname
    except AttributeError:
        pass
    params = []
    for name in _settable_fields(dataclass).keys():
        params.append(f"    :param {name}: Optionally initialize `{name}` field.\n")
    docstring = rf"""Builder for the :class:`{dname}` dataclass.

    This class allows the :class:`{dname}` dataclass to be constructed with the
    builder pattern.  Once an instance is constructed simply assign to it's
    attributes, which are identical to the :class:`{dname}` dataclass.  When
    done use it's `build` method, or the :func:`build` function if one of the
    fields is `build`, to make an instance of the :class:`{dname}` dataclass
    using the field values set on this builder.

    .. warning::

        Because this class overrides attribute assignment, care must be taken
        when extending to only use private and/or "dunder" attributes and
        methods.

    See :class:`{dname}` for further information on each filed.

{''.join(params)}

    :raises dataclass_builder.exceptions.UndefinedFieldError:
        If you try to assign to a field that is not part of :class:`{dname}`\ 's
        `__init__` method.
    :raises dataclass_builder.exceptions.MissingFieldError:
        If :func:`build` is called on this builder before all non default
        fields of the dataclass are assigned.
        """
    return docstring


def dataclass_builder(  # noqa: C901
    dataclass: Type[Any], *, name: Optional[str] = None
) -> Type[Any]:
    """Create a new builder class specialized to a given dataclass.

    :param dataclass:
        The :func:`dataclasses.dataclass` to create the builder for.
    :param name:
        Override the name of the builder, by default it will be
        '<dataclass>Builder' where <dataclass> is replaced by the name of the
        dataclass.

    :return object:
        A new dataclass builder class that is specialized to the given
        `dataclass`.  If the given :func:`dataclasses.dataclass` does not
        contain the fields `build` or `fields` these will be exposed as public
        methods with the same signature as the
        :func:`dataclass_builder.utility.build` and
        :func:`dataclass_builder.utility.fields` functions respectively.

    :raises TypeError:
        If `dataclass` is not a :func:`dataclasses.dataclass`. This is decided
        via :func:`dataclasses.is_dataclass`.
    """
    if not is_dataclass(dataclass):
        raise TypeError("must be called with a dataclass type")

    settable_fields = _settable_fields(dataclass)
    required_fields = _required_fields(dataclass)
    optional_fields = _optional_fields(dataclass)

    # validate identifiers
    for name_ in _settable_fields(dataclass):
        # there should not be anyway to trigger this branch
        if not name_.isidentifier():  # pragma: no cover
            raise RuntimeError(
                f"field name '{name_}'' could cause a security issue, refusing"
                f" to construct builder for '{dataclass.__qualname__}'"
            )

    dname = dataclass.__qualname__
    try:
        dname = dataclass.__module__ + "." + dname
    except AttributeError:
        pass

    def _setattr_method(self: Any, name: str, value: Any) -> None:
        # self.__initialized is not protected member access, since this is
        # a class method
        if (
            name.startswith("_") or hasattr(self, name) or not self.__initialized
        ):  # pylint: disable=protected-access
            object.__setattr__(self, name, value)
        else:
            raise UndefinedFieldError(
                f"dataclass '{dataclass.__qualname__}' does not define "
                f"field '{name}'",
                dataclass,
                name,
            )

    _setattr_method.__doc__ = f"""\
    Set a field value, or an object attribute if it is private.

        .. note::

            This will pass through all attributes beginning with an underscore.
            If this is a valid field of the dataclass it will still be built
            correctly but UndefinedFieldError will not be thrown for attributes
            beginning with an underscore.

            If you need the exception to be thrown then set the field in the
            constructor.

        :param name:
            Name of the dataclass field or private/dunder attribute to set.
        :param value:
            Value to assign to the dataclass field or private/dunder
            attribute.

        :raises dataclass_builder.exceptions.UndefinedFieldError:
            If `name` is not initialisable in the :class:`{dname}` dataclass.
            If `name` is private (begins with an underscore) or is a "dunder"
            then this exception will not be raised.
        """

    def _repr_method(self: Any) -> str:
        """Print a representation of the builder.

        >>> PointBuilder = dataclass_builder(Point)
        >>> PointBuilder(x=4.0, w=2.0)
        PointBuilder(x=4.0, w=2.0)

        :return:
            String representation that can be used to construct this builder
            instance.
        """
        args = []
        for name in settable_fields:
            value = getattr(self, name)
            if value not in (REQUIRED, OPTIONAL):
                args.append(f"{name}={repr(value)}")
        return f'{self.__class__.__qualname__}({", ".join(args)})'

    def _build_method(self: Any) -> Any:
        # check for missing required fields
        for name, field in required_fields.items():
            if getattr(self, name) is REQUIRED:
                raise MissingFieldError(
                    f"field '{name}' of dataclass '{dataclass.__qualname__}' "
                    "is not optional",
                    dataclass,
                    field,
                )
        # build dataclass
        kwargs = {
            name: getattr(self, name)
            for name in settable_fields
            if getattr(self, name) is not OPTIONAL
        }
        return dataclass(**kwargs)

    _build_method.__doc__ = f"""\
    Build a :class:`{dname}` dataclass using the fields from this builder.

    :return:
        An instance of the :class:`{dname}` dataclass using the fields set on
        this builder instance.

    :raises dataclass_builder.exceptions.MissingFieldError:
        If not all of the required fields have been assigned to this
        builder instance.
    """

    def _fields_method(
        _: Any, required: bool = True, optional: bool = True
    ) -> Mapping[str, "Field[Any]"]:
        if not required and not optional:
            return {}
        if required and not optional:
            return required_fields
        if not required and optional:
            return optional_fields
        return settable_fields

    _fields_method.__doc__ = f"""Get a dictionary of the builder's fields.

        :param required:
            Set to False to not report required fields.
        :param optional:
            Set to False to not report optional fields.

        :return:
            A mapping from field names to actual :class:`dataclasses.Field`'s
            in the same order as in the :class:`{dname}` dataclass.
        """

    # Fix return type of build, it won't help Mypy as it cannot handle
    # classes created at runtime but typing.get_type_hints will work properly.
    #
    # See: https://github.com/python/mypy/wiki/Unsupported-Python-Features
    _build_method.__annotations__["return"] = dataclass

    # assemble new builder class methods
    dict_: Dict[str, Any] = dict()
    dict_["__init__"] = _create_init_method(settable_fields)
    dict_["__setattr__"] = _setattr_method
    dict_["__repr__"] = _repr_method
    dict_["_build"] = _build_method
    dict_["_fields"] = _fields_method
    dict_["__doc__"] = _create_class_docstring(dataclass)

    if "build" not in settable_fields:
        dict_["build"] = _build_method

    if "fields" not in settable_fields:
        dict_["fields"] = _fields_method

    if name is None:
        name = f"{dataclass.__name__}Builder"

    return type(name, (object,), dict_)
